### DISCUSSION TITLE
Quote with children… but has no children?

### DISCUSSION THREAD

**Discussion by scrooks** [jheddings/notional#215] [Q&A]
I'm getting back a Quote that looks like this:

```
type='quote' object='block' id=UUID('3dd16b64-6ef5-4288-bfb0-d25d81194f02') parent=PageRef(type='page_id', page_id=UUID('88fb748b-3724-4ed8-ac4d-c60558ec34be')) has_children=True archived=False created_time=datetime.datetime(2023, 5, 10, 23, 11, tzinfo=datetime.timezone.utc) created_by=User(object='user', id=UUID('69172dd4-55e6-409f-be0a-e601bcce5323'), type=None, name=None, avatar_url=None) last_edited_time=datetime.datetime(2023, 5, 13, 22, 19, tzinfo=datetime.timezone.utc) last_edited_by=User(object='user', id=UUID('69172dd4-55e6-409f-be0a-e601bcce5323'), type=None, name=None, avatar_url=None) quote=_NestedData(rich_text=[TextObject(type='text', plain_text='Front Gate Area', href=None, annotations=Annotations(bold=True, italic=False, strikethrough=False, underline=False, code=False, color=<FullColor.DEFAULT: 'default'>), text=_NestedData(content='Front Gate Area', link=None))], children=None, color=<FullColor.DEFAULT: 'default'>)
```

Note that `has_children=True`, but `quote.children=None`.  I'm missing out on the items inside the Quote.  Am I doing something wrong?  Is this a bug?  I can't find a way to get to the text that is a child of that Quote.

**Comment by jheddings** [comment]
I will take a closer look, but most of the blocks in Notion do not return the children directly.  Instead, when `has_children` is True, you need to use the `notional.blocks.children.list` endpoint to get each of the child blocks.  I believe there is an example in the `page.py` script to demonstrate this.(I’m away from my computer to verify at the moment, but hopefully this gives you a start)-jhOn May 13, 2023, at 19:22, Steve Crooks ***@***.***> wrote:﻿
I'm getting back a Quote that looks like this:
type='quote' object='block' id=UUID('3dd16b64-6ef5-4288-bfb0-d25d81194f02') parent=PageRef(type='page_id', page_id=UUID('88fb748b-3724-4ed8-ac4d-c60558ec34be')) has_children=True archived=False created_time=datetime.datetime(2023, 5, 10, 23, 11, tzinfo=datetime.timezone.utc) created_by=User(object='user', id=UUID('69172dd4-55e6-409f-be0a-e601bcce5323'), type=None, name=None, avatar_url=None) last_edited_time=datetime.datetime(2023, 5, 13, 22, 19, tzinfo=datetime.timezone.utc) last_edited_by=User(object='user', id=UUID('69172dd4-55e6-409f-be0a-e601bcce5323'), type=None, name=None, avatar_url=None) quote=_NestedData(rich_text=[TextObject(type='text', plain_text='Front Gate Area', href=None, annotations=Annotations(bold=True, italic=False, strikethrough=False, underline=False, code=False, color=<FullColor.DEFAULT: 'default'>), text=_NestedData(content='Front Gate Area', link=None))], children=None, color=<FullColor.DEFAULT: 'default'>)

Note that has_children=True, but quote.children=None.  I'm missing out on the items inside the Quote.  Am I doing something wrong?  Is this a bug?  I can't find a way to get to the text that is a child of that Quote.

**Reply by scrooks** [reply]
The problem in this case is that `notional.block.children` is `None` even though `has_children` is `True`.  I was going to "follow the children", but they seem to have not arrived.  That's why I wondered if this was a bug somewhere, or if I was approaching it wrong.

**Comment by jheddings** [comment]
I may have misunderstood your use case…  Are you just trying to get the text contained in a quote block, or do you have a quote with additional blocks as children?If you have additional child blocks, you will need to make another API call to get the children of the quote block.  The `children` element in responses is almost always `None` (which is unfortunate).On May 19, 2023, at 21:43, Steve Crooks ***@***.***> wrote:﻿
The problem in this case is that notional.block.children is None even though has_children is True.  I was going to "follow the children", but they seem to have not arrived.  That's why I wondered if this was a bug somewhere, or if I was approaching it wrong.

**Reply by scrooks** [reply]
Yes, that's right, I'm trying to get additional child blocks.  If the `children` element is `None`, what information should be used to make another call to get the children?

**Comment by scrooks** [comment]
I think I figured this out, but I don't know for sure that I did it the "correct" way.

Here's the code I use, which recursively pulls in child blocks.  I'm trying to get everything back in Markdown format, so I special case Image and Toggle blocks since they don't have a `Markdown` property.

```python
def pull_content(notional_item):
    notion = notional_item._notional__session
    page = notion.pages.retrieve(notional_item.id)
    with io.StringIO() as f:
        _pull_content_blocks(f, notion, page)
        content = f.getvalue().strip()
        return content


def _pull_content_blocks(f, notion, page, prefix=""):
    for block in notion.blocks.children.list(page):
        child_prefix = prefix
        if isinstance(block, Image):
            f.write(f"{prefix}![]({block.image.URL})\n\n")
        elif isinstance(block, Toggle):
            value = block.PlainText
            if block.toggle.rich_text[0].annotations.italic:
                value = f"_{value}_"
            if block.toggle.rich_text[0].annotations.bold:
                value = f"**{value}**"
            child_prefix = f"> {prefix}"
            f.write(f"{child_prefix}{value}\n\n")
        else:
            f.write(f"{prefix}{block.Markdown}\n\n")
        if block.has_children:
            _pull_content_blocks(f, notion, block, child_prefix)
